### PR TITLE
Fix curve spell and use registry in oracle

### PR DESCRIPTION
**Pull request by n1punp**

### CODE REVIEW

File: contracts/oracle/CurveOracle.sol [hunk 1]
```diff
@@ -13,26 +14,48 @@ contract CurveOracle is IBaseOracle {
...
+    registry = _registry;
+  }
+
+  /// @dev Return pool address given LP token and update pool info if not exist.
+  /// @param lp LP token to find the corresponding pool.
+  function getPool(address lp) public returns (address) {
+    address pool = poolOf[lp];
+    if (pool == address(0)) {
+      require(lp != address(0), 'no lp token');
+      pool = registry.get_pool_from_lp_token(lp);
```
Comment (GriphookETH) [inline contracts/oracle/CurveOracle.sol:33]: Will this call revert if the pool does not exist?

File: contracts/oracle/CurveOracle.sol [hunk 2]
```diff
@@ -13,26 +14,48 @@ contract CurveOracle is IBaseOracle {
...
-        if (decimals > 18) tokenPx = tokenPx.mul(10**(decimals - 18));
-        if (tokenPx < minPx) minPx = tokenPx;
-      } catch (bytes memory) {
-        break;
-      }
+    address[] memory tokens = ulTokens[lp];
+    uint n = tokens.length;
+    for (uint idx = 0; idx < n; idx++) {
+      address token = tokens[idx];
+      uint decimals = IERC20Decimal(token).decimals();
```
Comment (GriphookETH) [inline contracts/oracle/CurveOracle.sol:54]: If we are doing the catching, I think we can do another step and also catch decimals.
```
struct UnderlyingToken {
   uint8 decimals;
   address token;
}
```
^ this fits into a single word. It won't work with token that decimals can change, but I don't think those exist.

**Review by GriphookETH — CHANGES_REQUESTED** [review]

File: contracts/oracle/CurveOracle.sol (hunk 1 above)
Comment (n1punp) [reply contracts/oracle/CurveOracle.sol:38]: Default is zero address. Added require.

File: contracts/oracle/CurveOracle.sol [hunk 3]
```diff
@@ -13,26 +14,54 @@ contract CurveOracle is IBaseOracle {
...
+    if (pool == address(0)) {
+      require(lp != address(0), 'no lp token');
+      pool = registry.get_pool_from_lp_token(lp);
+      require(pool != address(0), 'no corresponding pool for lp token');
+      poolOf[lp] = pool;
+      uint n = registry.get_n_coins(pool);
+      address[8] memory tokens = registry.get_coins(pool);
+      for (uint i = 0; i < n; i++) {
+        ulTokens[lp].push(
+          UnderlyingToken({token: tokens[i], decimals: uint8(IERC20Decimal(tokens[i]).decimals())})
```
Comment (GriphookETH) [inline contracts/oracle/CurveOracle.sol:45]: `uint8` casting should not be necessary

**Review by GriphookETH — CHANGES_REQUESTED** [review]

**Review by GriphookETH — APPROVED** [review]

**Pull request merged**